README.md
# Kodiアドオン：フォトビューア

macOSの写真アプリケーションで管理されている写真をTV画面で閲覧するためのKodiアドオンです。
写真アプリケーションのデータベースを参照し、撮影日、撮影場所、顔画像認識による人物等にしたがって写真やビデオをブラウズできます。

macOS High Sierra（バージョン10.13）の写真アプリケーション（バージョン3.0）にて動作を確認しています。
最新の写真アプリケーションではデータベースの構成が異なるため動作しません。

はじめに[アドオン設定画面-設定](アドオン設定画面-設定)で写真ライブラリへのパスを設定してください。


## メイン画面

アドオン起動直後にメイン画面が表示されます。

![メイン画面](https://github.com/kodiful/plugin.image.photosviewer/assets/12268536/8c0ee9d1-a994-455a-83ce-cc50871ddc68)

### 写真

写真の撮影日にしたがって年月日別にブラウズします。

![撮影日ブラウズ画面](https://github.com/kodiful/plugin.image.photosviewer/assets/12268536/8c0ee9d1-a994-455a-83ce-cc50871ddc68)

![撮影日ブラウズ画面](https://github.com/kodiful/plugin.image.photosviewer/assets/12268536/66712300-154e-4906-a6c4-7755ac820ae7)

![撮影日ブラウズ画面](https://github.com/kodiful/plugin.image.photosviewer/assets/12268536/4e4bcf7d-2ea8-469b-a4e1-922f680eb99c)

![撮影日ブラウズ画面](https://github.com/kodiful/plugin.image.photosviewer/assets/12268536/852cfa78-bb75-4a0b-a7a9-d459b20b45b1)

![撮影日ブラウズ画面](https://github.com/kodiful/plugin.image.photosviewer/assets/12268536/b3edaaca-6d29-4653-8f67-5992dbe22fcd)

![撮影日ブラウズ画面](https://github.com/kodiful/plugin.image.photosviewer/assets/12268536/7ceed763-4062-4b73-8dc1-853ca69a1d07)

### 写真のコンテクストメニュー

撮影日の絞り込み時に、年または月単位にまとめてブラウズできます。

![撮影日ブラウズ画面コンテクストメニュー](https://github.com/kodiful/plugin.image.photosviewer/assets/12268536/7ef431e3-6711-4660-8d1c-206332816cc6)

### ピープル

写真からの顔画像認識にしたがって人物別にブラウズします。

### 撮影地

写真に設定されている位置情報にしたがって撮影地別にブラウズします。

![撮影地ブラウズ画面](https://github.com/kodiful/plugin.image.photosviewer/assets/12268536/f86e359c-2505-4745-b0c7-f41542aa5edb)

![撮影地ブラウズ画面](https://github.com/kodiful/plugin.image.photosviewer/assets/12268536/bb6d2d73-0144-4559-8fe0-bc44e41351ab)

![撮影地ブラウズ画面](https://github.com/kodiful/plugin.image.photosviewer/assets/12268536/0a20f8b1-2516-400d-b72e-2a969a8e2c7a)

![撮影地ブラウズ画面](https://github.com/kodiful/plugin.image.photosviewer/assets/12268536/d09e4da1-c7ad-4072-9478-c739aaaaa347)

![撮影地ブラウズ画面](https://github.com/kodiful/plugin.image.photosviewer/assets/12268536/a88dc006-37fc-41fd-87a6-ea9f0c8ae7fc)

![撮影地ブラウズ画面](https://github.com/kodiful/plugin.image.photosviewer/assets/12268536/34eef5ff-67cd-44da-9dc0-2b9324e3ea17)


### 撮影地のコンテクストメニュー

撮影地の絞り込み時に、都道府県や市町村等の地域単位にまとめてブラウズできます。

![撮影地ブラウズ画面コンテクストメニュー](https://github.com/kodiful/plugin.image.photosviewer/assets/12268536/240262a9-bace-4685-bdab-4e222eb105ee)

### ビデオ

ビデオをブラウズします。

![ビデオブラウズ画面](https://github.com/kodiful/plugin.image.photosviewer/assets/12268536/7cd74462-e9f5-4ae4-92c6-f13af8d5b7ca)

![video-select](https://github.com/kodiful/plugin.image.photosviewer/assets/12268536/a27bbcbd-e193-4b2d-b562-7533dde8702f)

### アルバム

写真アプリケーションのアルバムの階層にしたがってブラウズします。

![アルバムブラウズ画面](https://github.com/kodiful/plugin.image.photosviewer/assets/12268536/e1524cb4-1410-4636-b1e5-b12d60de1751)

![アルバムブラウズ画面](https://github.com/kodiful/plugin.image.photosviewer/assets/12268536/bc55b9af-5714-46e4-bebe-de4de11cba9a)


## アドオン設定画面-設定

![アドオン設定画面](https://github.com/kodiful/plugin.image.photosviewer/assets/12268536/add5c124-ce85-4c90-ba72-526b3e26c4f8)

### 写真ライブラリへのパス

写真ライブラリへのパス（デフォルトでは（"/Users/(username)/Pictures/写真ライブラリ.photoslibrary"）を設定してください。


## アドオン設定画面-その他

![アドオン設定画面](https://github.com/kodiful/plugin.image.photosviewer/assets/12268536/cd682ccd-e414-476d-9dda-249a34114c0c)

### HEIC画像の表示

iOS11から採用された写真データの形式HEICは、そのままではKodiで表示できないため、これを表示する方法を「サムネイルで代替」「JPEGに変換」から選択してください。それぞれ以下の特徴があります。

|方法|特徴|
|:---|:---|
|サムネイルで代替|サムネイル画像で置き換えます。表示に時間はかかりませんが、画質はよくありません。|
|JPEGに変換|macOSのsipsコマンドによりHEIC画像をJPEG画像に変換して表示します。変換に時間を要しますが、高画質です。|
|変換しない|HEIF image decoderが利用可能な場合はこれを選択してください。|

### デバッグ

通常はオフにしてください。




CHANGELOG.txt
2021-02-23 v2.0
Kodi 19 Matrixに対応

2020-10-18 v1.4.1
祝祭日のデータを更新

2019-05-07 v1.4
HEIF image decoderが利用可能な場合に対応して、HEIC画像の表示として変換しない設定を追加

2019-05-07 v1.3.1
祝祭日のデータを更新

2018-05-04 v1.3
撮影地を地理情報の木構造にしたがって表示するよう変更
削除した画像は表示から除外するよう変更

2018-04-30 v1.2
アイコン、サムネイルをicon8の画像に差し替え
祝祭日のデータを更新

2018-03-13 v1.1
HEIC画像をJPEG変換する機能を追加

2018-03-12 v1.0.3
HEIC画像が表示されないためサムネイルのJPEG画像で代替

2018-02-15 v1.0.2
icon.pngを変更

2017−12−23 v1.0.1
addon.xmlにwebsiteを追記

2017-05-01 v1.0
リリース


resources/lib/db.py
# -*- coding: utf-8 -*-

from sqlite3 import dbapi2 as sqlite

from resources.lib.common import log


class DB:

    def __init__(self, dbfile):
        self.OpenDB(dbfile)

    def __del__(self):
        self.CloseDB()

    def OpenDB(self, dbfile):
        try:
            self.dbconn = sqlite.connect(dbfile)
        except Exception as e:
            log('db error: %s' % e, error=True)

    def CloseDB(self):
        try:
            self.dbconn.close()
        except Exception as e:
            log('db error: %s' % e, error=True)

    def GetMomentList(self, year, month):
        moment_list = []
        cur = self.dbconn.cursor()
        try:
            if year is None:
                cur.execute('''SELECT strftime('%Y', imageDate+978307200, 'unixepoch', 'localtime') as year
                               FROM RKMaster m
                               WHERE m.isInTrash = 0
                               GROUP BY year
                               ORDER BY imageDate
                               ''')
            elif month is None:
                cur.execute('''SELECT strftime('%m', imageDate+978307200, 'unixepoch', 'localtime') as month
                               FROM RKMaster m
                               WHERE m.isInTrash = 0
                               AND strftime('%Y', imageDate+978307200, 'unixepoch', 'localtime') = ?
                               GROUP BY month
                               ORDER BY imageDate
                               ''', ('%s' % (year),))
            else:
                cur.execute('''SELECT strftime('%d', imageDate+978307200, 'unixepoch', 'localtime') as day
                               FROM RKMaster m
                               WHERE m.isInTrash = 0
                               AND strftime('%Y-%m', imageDate+978307200, 'unixepoch', 'localtime') = ?
                               GROUP BY day
                               ORDER BY imageDate
                               ''', ('%s-%s' % (year, month),))
            for row in cur:
                moment_list.append(row)
        except Exception as e:
            log('db error: %s' % e, error=True)
        cur.close()
        return moment_list

    def GetPlaceList(self, parent=None):
        place_list = []
        cur = self.dbconn.cursor()
        try:
            if parent is None:
                cur.execute('''SELECT defaultName, uuid, type, modelId, minLatitude, maxLatitude, minLongitude, maxLongitude FROM RKPlace
                               WHERE type = 1
                               ORDER BY defaultName ASC''')
            else:
                (name, uuid, type, modelId) = parent
                if type == 1:
                    to = 2
                elif type == 2:
                    type = 4
                    to = 16
                else:
                    to = 100
                cur.execute('''SELECT DISTINCT p.defaultName, p.uuid, p.type, p.modelId, p.minLatitude, p.maxLatitude, p.minLongitude, p.maxLongitude FROM RKPlaceForVersion pv, RKPlace p
                               WHERE pv.versionId IN (SELECT versionId FROM RKPlaceForVersion WHERE placeId=?)
                               AND p.modelId = pv.placeId
                               AND p.type > ? AND p.type <= ?
                               ORDER BY p.minLatitude+p.maxLatitude+p.minLatitude+p.maxLongitude DESC''', (modelId, type, to,))
            for row in cur:
                place_list.append(row)
        except Exception as e:
            log('db error: %s' % e, error=True)
        cur.close()
        return place_list

    def GetPersonList(self):
        person_list = []
        cur = self.dbconn.cursor()
        try:
            cur.execute('''SELECT p.name, p.uuid, m.modelId
                           FROM RKModelResource m, RKPerson p
                           WHERE m.attachedModelId = p.representativeFaceId
                           AND p.name != ""
                           ORDER BY p.name ASC''')
            for row in cur:
                person_list.append(row)
        except Exception as e:
            log('db error: %s' % e, error=True)
        cur.close()
        return person_list

    def GetFolderList(self, folderUuid):
        folder_list = []
        cur = self.dbconn.cursor()
        try:
            cur.execute('''SELECT f.name, f.uuid
                           FROM RKFolder f
                           WHERE f.isHidden = 0 AND f.isInTrash = 0 AND f.parentFolderUuid = ?
                           ORDER BY f.name ASC''', (folderUuid,))
            for row in cur:
                folder_list.append(row)
        except Exception as e:
            log('db error: %s' % e, error=True)
        cur.close()
        return folder_list

    def GetAlbumList(self, folderUuid):
        album_list = []
        cur = self.dbconn.cursor()
        try:
            cur.execute('''SELECT a.name, a.uuid, v.modelId
                           FROM RKVersion v, RKAlbum a
                           WHERE v.uuid = a.posterVersionUuid
                           AND a.isHidden = 0 AND a.isInTrash = 0 AND a.customSortAvailable = 1 AND a.folderUuid = ?
                           ORDER BY a.name ASC''', (folderUuid,))
            for row in cur:
                album_list.append(row)
        except Exception as e:
            log('db error: %s' % e, error=True)
        cur.close()
        return album_list

    def GetVideoList(self):
        video_list = []
        cur = self.dbconn.cursor()
        try:
            cur.execute('''SELECT m.imageDate, m.imagePath, m.isMissing, v.modelId, v.latitude, v.longitude
                           FROM RKMaster m, RKVersion v
                           WHERE m.uuid = v.masterUuid
                           AND m.isInTrash = 0 AND m.isMissing = 0 AND v.naturalDuration > 0
                           ORDER BY m.imageDate ASC''')
            for row in cur:
                video_list.append(row)
        except Exception as e:
            log('db error: %s' % e, error=True)
        cur.close()
        return video_list

    def GetPictureList(self, uuid, action):
        picture_list = []
        cur = self.dbconn.cursor()
        try:
            if action == 'moments':
                cur.execute('''SELECT m.imageDate, m.imagePath, m.isMissing, v.modelId, v.latitude, v.longitude, v.orientation
                               FROM RKMaster m, RKVersion v
                               WHERE m.uuid = v.masterUuid
                               AND m.isInTrash = 0
                               AND m.uuid = ?
                               GROUP BY m.uuid
                               ORDER BY m.imageDate ASC''', (uuid,))
            elif action == 'people':
                cur.execute('''SELECT m.imageDate, m.imagePath, m.isMissing, v.modelId, v.latitude, v.longitude, v.orientation
                               FROM RKMaster m, RKVersion v, RKFace f, RKPerson p
                               WHERE m.uuid = v.masterUuid
                               AND m.isInTrash = 0
                               AND v.modelId = f.imageModelId
                               AND f.personId = p.modelId
                               AND p.uuid = ?
                               GROUP BY m.uuid
                               ORDER BY m.imageDate ASC''', (uuid,))
            elif action == 'places':
                cur.execute('''SELECT m.imageDate, m.imagePath, m.isMissing, v.modelId, v.latitude, v.longitude, v.orientation
                               FROM RKMaster m, RKVersion v, RKPlace p, RKPLaceForVersion pv
                               WHERE m.uuid = v.masterUuid
                               AND m.isInTrash = 0
                               AND v.modelId = pv.versionId
                               AND pv.placeId = p.modelId
                               AND p.uuid = ?
                               GROUP BY m.uuid
                               ORDER BY m.imageDate ASC''', (uuid,))
            elif action == 'search_by_year':
                (year) = uuid
                cur.execute('''SELECT m.imageDate, m.imagePath, m.isMissing, v.modelId, v.latitude, v.longitude, v.orientation
                               FROM RKMaster m, RKVersion v
                               WHERE m.uuid = v.masterUuid
                               AND m.isInTrash = 0
                               AND strftime('%Y', m.imageDate+978307200, 'unixepoch', 'localtime') = ?
                               GROUP BY m.uuid
                               ORDER BY m.imageDate ASC''', ('%s' % (year),))
            elif action == 'search_by_month':
                (year, month) = uuid
                cur.execute('''SELECT m.imageDate, m.imagePath, m.isMissing, v.modelId, v.latitude, v.longitude, v.orientation
                               FROM RKMaster m, RKVersion v
                               WHERE m.uuid = v.masterUuid
                               AND m.isInTrash = 0
                               AND strftime('%Y-%m', m.imageDate+978307200, 'unixepoch', 'localtime') = ?
                               GROUP BY m.uuid
                               ORDER BY m.imageDate ASC''', ('%s-%s' % (year, month),))
            elif action == 'search_by_day':
                (year, month, day) = uuid
                cur.execute('''SELECT m.imageDate, m.imagePath, m.isMissing, v.modelId, v.latitude, v.longitude, v.orientation
                               FROM RKMaster m, RKVersion v
                               WHERE m.uuid = v.masterUuid
                               AND m.isInTrash = 0
                               AND strftime('%Y-%m-%d', m.imageDate+978307200, 'unixepoch', 'localtime') = ?
                               GROUP BY m.uuid
                               ORDER BY m.imageDate ASC''', ('%s-%s-%s' % (year, month, day),))
            elif action == 'search_by_timestamp':
                (timestamp) = uuid
                cur1 = self.dbconn.cursor()
                cur1.execute('''SELECT strftime('%Y-%m-%d', ? + 978307200, 'unixepoch', 'localtime')''', ('%d' % (int(float(timestamp))),))
                (date) = cur1.fetchone()
                cur1.close()
                cur.execute('''SELECT m.imageDate, m.imagePath, m.isMissing, v.modelId, v.latitude, v.longitude, v.orientation
                               FROM RKMaster m, RKVersion v
                               WHERE m.uuid = v.masterUuid
                               AND m.isInTrash = 0
                               AND strftime('%Y-%m-%d', m.imageDate+978307200, 'unixepoch', 'localtime') = ?
                               GROUP BY m.uuid
                               ORDER BY m.imageDate ASC''', (date[0],))
            elif action == 'search_by_latlong':
                (latitude, longitude) = uuid
                cur.execute('''SELECT m.imageDate, m.imagePath, m.isMissing, v.modelId, v.latitude, v.longitude, v.orientation
                               FROM RKMaster m, RKVersion v
                               WHERE m.uuid = v.masterUuid
                               AND m.isInTrash = 0
                               AND v.latitude != ""
                               AND v.longitude != ""
                               GROUP BY m.uuid
                               ORDER BY abs(v.latitude - ?) + abs(v.longitude - ?) ASC LIMIT 100''', (latitude, longitude,))
            else:
                cur.execute('''SELECT m.imageDate, m.imagePath, m.isMissing, v.modelId, v.latitude, v.longitude, v.orientation
                               FROM RKMaster m, RKVersion v, RKCustomSortOrder o
                               WHERE m.uuid = v.masterUuid
                               AND m.isInTrash = 0
                               AND v.Uuid = o.objectUuid
                               AND o.containerUuid = ?
                               GROUP BY m.uuid
                               ORDER BY m.imageDate ASC''', (uuid,))
            for row in cur:
                picture_list.append(row)
        except Exception as e:
            log('db error: %s' % e, error=True)
        cur.close()
        return picture_list


resources/lib/map.py
# -*- coding: utf-8 -*-

import os

from sqlite3 import dbapi2 as sqlite
from urllib.request import urlretrieve
from urllib.request import urlcleanup
from urllib.error import HTTPError

import xbmc
import xbmcvfs

from resources.lib.const import Const


class Map:

    def __init__(self):
        # ディレクトリ
        self.path = os.path.join(Const.PROFILE_PATH, 'cache', 'thumbnail')
        if not os.path.isdir(self.path):
            os.makedirs(self.path)
        # DB
        self.db = os.path.join(xbmcvfs.translatePath('special://database'), 'Textures13.db')
        # APIキー
        self.key = Const.GET('apikey')
        # 言語
        self.lang = Const.STR(30020)
        # 地域
        self.region = Const.STR(30021)

    def create(self, minLatLong, maxLatLong=None):
        # クエリ作成 https://developers.google.com/maps/documentation/static-maps/intro
        coordinates = '%s,%s' % minLatLong
        if maxLatLong:
            coordinates = coordinates + '|%s,%s' % maxLatLong
        # ファイルパス
        filepath = os.path.join(self.path, '%s.png' % coordinates)
        # 不完全なファイルは削除
        if os.path.isfile(filepath) and os.path.getsize(filepath) < 12000:
            try:
                # delete imagefile
                os.remove(filepath)
                # delete from database
                conn = sqlite.connect(self.db)
                c = conn.cursor()
                c.execute("DELETE FROM texture WHERE url = ?", (filepath,))
                conn.commit()
                conn.close()
            except Exception:
                pass
        # ファイルを取得
        if not os.path.isfile(filepath):
            if self.key:
                # APIキーが設定されていたら画像ファイルを取得
                try:
                    try:
                        xbmc.sleep(1000)
                    except Exception:
                        pass
                    # リクエストパラメータ
                    params = (
                        ('key', self.key),
                        ('language', self.lang),
                        ('region', self.region),
                        ('visible', coordinates),
                        ('size', '320x320'),
                        ('scale', '2'),
                        ('maptype', 'roadmap'),
                        ('style', 'feature:road.local|element:geometry|visibility:simplified'),
                        ('style', 'feature:administrative|element:labels.text.fill|color:0x000000'),
                        ('style', 'feature:administrative.province|element:geometry.stroke|weight:2|color:0x444444'),
                    )
                    # リクエスト実行
                    urlretrieve(
                        'http://maps.googleapis.com/maps/api/staticmap?%s' % '&'.join(map(lambda x: '%s=%s' % x, params)),
                        filepath)
                except HTTPError as e:
                    raise e
                except Exception:
                    urlcleanup()
                    remove_tries = 3
                    while remove_tries and os.path.isfile(filepath):
                        try:
                            os.remove(filepath)
                        except Exception:
                            remove_tries -= 1
                        try:
                            xbmc.sleep(1000)
                        except Exception:
                            pass
        # 完全なファイルが取得できなかったらデフォルトの画像ファイルを設定
        if not os.path.isfile(filepath) or os.path.getsize(filepath) < 12000:
            filepath = Const.MARKER
        return filepath


resources/lib/cache.py
# -*- coding: utf-8 -*-

import os
import subprocess
import hashlib

from resources.lib.const import Const


class Cache:

    def __init__(self):
        # ディレクトリ
        self.path = os.path.join(Const.PROFILE_PATH, 'cache', 'heic2jpeg')
        # ディレクトリが無ければ作成
        if not os.path.isdir(self.path):
            os.makedirs(self.path)

    def clear(self):
        files = os.listdir(self.path)
        for filename in files:
            os.remove(os.path.join(self.path, filename))

    def convert(self, infile):
        # 出力ファイル
        outfile = os.path.join(self.path, '%s.jpeg' % (hashlib.md5(infile).hexdigest()))
        # 画像変換実行
        if not os.path.isfile(outfile):
            command = 'sips --setProperty format jpeg "{infile}" --out "{outfile}"'.format(infile=infile, outfile=outfile)
            subprocess.call(command, shell=True)
        return outfile


addon.py
# -*- coding: utf-8 -*-

import sys
import os
import glob
import datetime

from urllib.parse import parse_qs
from urllib.parse import urlencode

import xbmcgui
import xbmcplugin
import xbmcvfs

from resources.lib.common import notify
from resources.lib.common import isholiday
from resources.lib.db import DB
from resources.lib.map import Map
from resources.lib.cache import Cache
from resources.lib.const import Const


def build_url(values):
    query = urlencode(values)
    if query:
        url = '%s?%s' % (Const.ADDON_URL, query)
    else:
        url = Const.ADDON_URL
    return url


def convert_timestamp(year=None, month=None, day=None, hour=None, minute=None, timestamp=None):
    if timestamp:
        t = datetime.datetime.fromtimestamp(int(timestamp) + 978307200)
        (year, month, day, hour, minute) = t.replace(year=t.year).strftime('%Y,%m,%d,%H,%M').split(',')
    if year:
        if month:
            m = Const.STR(30017).split(',')
            mindex = int(month) - 1
            if day:
                d = Const.STR(30018).split(',')
                dindex = datetime.date(int(year), int(month), int(day)).weekday()
                if hour and minute:
                    itemname = Const.STR(30030).format(year=year, month=m[mindex], day=day, day7=d[dindex], hour=hour, minute=minute)
                else:
                    itemname = Const.STR(30031).format(year=year, month=m[mindex], day=day, day7=d[dindex])
                if isholiday('%s-%s-%s' % (year, month, day)) or dindex == 6:
                    itemname = '[COLOR red]' + itemname + '[/COLOR]'
                elif dindex == 5:
                    itemname = '[COLOR blue]' + itemname + '[/COLOR]'
            else:
                itemname = Const.STR(30032).format(year=year, month=m[mindex])
        else:
            itemname = Const.STR(30033).format(year=year)
    else:
        itemname = ''
    return itemname


class App:

    def __init__(self):
        # ハンドル
        self.handle = int(sys.argv[1])
        # 写真アプリのパス
        self.photo_app_path = Const.GET('photo_library_path')
        if self.photo_app_path == '':
            self.photo_app_path = os.path.join(os.getenv('HOME'), 'Pictures', Const.STR(30000))
        Const.SET('photo_library_path', self.photo_app_path)
        # photos.dbをチェック
        self.photo_app_db_file = os.path.join(Const.PROFILE_PATH, 'photos.db')
        self.photo_app_db_orig = os.path.join(self.photo_app_path, 'database', 'photos.db')
        ctime = os.stat(self.photo_app_db_orig).st_ctime
        mtime = os.stat(self.photo_app_db_orig).st_mtime
        # photos.dbをコピー
        if xbmcvfs.exists(self.photo_app_db_file) and os.stat(self.photo_app_db_file).st_mtime == mtime:
            pass
        else:
            xbmcvfs.copy(self.photo_app_db_orig, self.photo_app_db_file)
            os.utime(self.photo_app_db_file, (ctime, mtime))
        # ファイルのパス
        self.photo_app_picture_path = os.path.join(self.photo_app_path, 'Masters')
        self.photo_app_thumbnail_path = os.path.join(self.photo_app_path, 'resources', 'proxies', 'derivatives')
        self.photo_app_face_path = os.path.join(self.photo_app_path, 'resources', 'media', 'face')
        # DB
        self.db = None

    def open_db(self):
        if self.db is not None:
            return
        try:
            self.db = DB(self.photo_app_db_file)
        except Exception:
            pass
        return self

    def close_db(self):
        try:
            self.db.CloseDB()
        except Exception:
            pass
        return self

    def list_moments(self, year, month):
        n = 0
        moments = self.db.GetMomentList(year, month)
        for (name,) in moments:
            if year is None:
                url = build_url({'action': 'moments', 'year': name})
                item = xbmcgui.ListItem(convert_timestamp(year=name))
                item.setArt({'icon': Const.CALENDAR, 'thumb': Const.CALENDAR})
                contextmenu = []
                contextmenu.append((Const.STR(30012).format(period=convert_timestamp(year=name)), 'Container.Update(%s)' % build_url({'action': 'search_by_year', 'year': name})))
                item.addContextMenuItems(contextmenu, replaceItems=True)
            elif month is None:
                url = build_url({'action': 'moments', 'year': year, 'month': name})
                item = xbmcgui.ListItem(convert_timestamp(year=year, month=name))
                item.setArt({'icon': Const.CALENDAR, 'thumb': Const.CALENDAR})
                contextmenu = []
                contextmenu.append((Const.STR(30012).format(period=convert_timestamp(year=year, month=name)), 'Container.Update(%s)' % build_url({'action': 'search_by_month', 'year': year, 'month': name})))
                contextmenu.append((Const.STR(30012).format(period=convert_timestamp(year=year)), 'Container.Update(%s)' % build_url({'action': 'search_by_year', 'year': year})))
                item.addContextMenuItems(contextmenu, replaceItems=True)
            else:
                url = build_url({'action': 'search_by_day', 'year': year, 'month': month, 'day': name})
                item = xbmcgui.ListItem(convert_timestamp(year=year, month=month, day=name))
                item.setArt({'icon': Const.CALENDAR, 'thumb': Const.CALENDAR})
                contextmenu = []
                contextmenu.append((Const.STR(30012).format(period=convert_timestamp(year=year, month=month)), 'Container.Update(%s)' % build_url({'action': 'search_by_month', 'year': year, 'month': month})))
                contextmenu.append((Const.STR(30012).format(period=convert_timestamp(year=year)), 'Container.Update(%s)' % build_url({'action': 'search_by_year', 'year': year})))
                item.addContextMenuItems(contextmenu, replaceItems=True)
            xbmcplugin.addDirectoryItem(self.handle, url, item, True)
            n += 1
        return n

    def list_places(self, parent=None):
        n = 0
        map = Map()
        places = self.db.GetPlaceList(parent)
        for (name, uuid, type, modelId, minLatitude, maxLatitude, minLongitude, maxLongitude) in places:
            try:
                thumbnail = map.create((minLatitude, minLongitude), (maxLatitude, maxLongitude))
            except Exception:
                thumbnail = Const.MARKER
            if int(type) > 16:
                url = build_url({'action': 'places', 'uuid': uuid})
                item = xbmcgui.ListItem(name)
                item.setArt({'icon': thumbnail, 'thumb': thumbnail})
            else:
                url = build_url({'action': 'places', 'name': name, 'uuid': uuid, 'type': type, 'modelId': modelId})
                item = xbmcgui.ListItem(name)
                item.setArt({'icon': thumbnail, 'thumb': thumbnail})
                contextmenu = []
                contextmenu.append((Const.STR(30013).format(place=name), 'Container.Update(%s)' % build_url({'action': 'places', 'uuid': uuid})))
                item.addContextMenuItems(contextmenu, replaceItems=True)
            xbmcplugin.addDirectoryItem(self.handle, url, item, True)
            n += 1
        return n

    def list_people(self):
        n = 0
        people = self.db.GetPersonList()
        for (name, uuid, modelId) in people:
            url = build_url({'action': 'people', 'uuid': uuid})
            imagePath = glob.glob(os.path.join(self.photo_app_face_path, ('%04x' % modelId)[0:2], '00', 'facetile_%x.jpeg' % modelId))[-1]
            item = xbmcgui.ListItem(name)
            item.setArt({'icon': imagePath, 'thumb': imagePath})
            xbmcplugin.addDirectoryItem(self.handle, url, item, True)
            n += 1
        return n

    def list_albums(self, folderUuid):
        n = 0
        folders = self.db.GetFolderList(folderUuid)
        for (name, uuid) in folders:
            url = build_url({'action': 'albums', 'folderUuid': uuid})
            thumbnailPath = Const.PHOTO_GALLERY
            item = xbmcgui.ListItem(name)
            item.setArt({'icon': thumbnailPath, 'thumb': thumbnailPath})
            xbmcplugin.addDirectoryItem(self.handle, url, item, True)
            n += 1
        albums = self.db.GetAlbumList(folderUuid)
        for (name, uuid, modelId) in albums:
            url = build_url({'action': 'albums', 'uuid': uuid})
            thumbnailPath = glob.glob(os.path.join(self.photo_app_thumbnail_path, ('%04x' % modelId)[0:2], '00', '%x' % modelId, '*.jpg'))[-1]
            item = xbmcgui.ListItem(name)
            item.setArt({'icon': thumbnailPath, 'thumb': thumbnailPath})
            xbmcplugin.addDirectoryItem(self.handle, url, item, True)
            n += 1
        return n

    def list_photos(self, uuid, action):
        pictures = self.db.GetPictureList(uuid, action)
        heic = Const.GET('heic')
        n = 0
        for (imageDate, imagePath, isMissing, modelId, latitude, longitude, orientation) in pictures:
            thumbnailPath = glob.glob(os.path.join(self.photo_app_thumbnail_path, ('%04x' % modelId)[0:2], '00', '%x' % modelId, '*.jpg'))[-1]
            if isMissing == 0:
                imagePath = os.path.join(self.photo_app_picture_path, imagePath)
            else:
                imagePath = thumbnailPath
            # replace heic images with thumbnails
            if imagePath.endswith('.HEIC') or imagePath.endswith('.heic'):
                if heic == '0':
                    # サムネイルで代替
                    imagePath = thumbnailPath
                elif heic == '1':
                    # JPEGに変換
                    imagePath = Cache().convert(imagePath)
                else:
                    pass
            item = xbmcgui.ListItem(convert_timestamp(timestamp=imageDate))
            item.setArt({'icon': thumbnailPath, 'thumb': thumbnailPath})
            contextmenu = []
            contextmenu.append((Const.STR(30010), 'Container.Update(%s)' % build_url({'action': 'search_by_timestamp', 'timestamp': imageDate})))
            if latitude and longitude:
                contextmenu.append((Const.STR(30011), 'Container.Update(%s)' % build_url({'action': 'search_by_latlong', 'latitude': latitude, 'longitude': longitude})))
            contextmenu.append((Const.STR(30014), 'Container.Update(%s,replace)' % build_url({})))
            item.addContextMenuItems(contextmenu, replaceItems=True)
            xbmcplugin.addDirectoryItem(self.handle, imagePath, item, False)
            n += 1
        return n

    def list_videos(self):
        n = 0
        videos = self.db.GetVideoList()
        for (imageDate, imagePath, isMissing, modelId, latitude, longitude) in videos:
            thumbnailPath = glob.glob(os.path.join(self.photo_app_thumbnail_path, ('%04x' % modelId)[0:2], '00', '%x' % modelId, '*.jpg'))[-1]
            if isMissing == 0:
                imagePath = os.path.join(self.photo_app_picture_path, imagePath)
            else:
                imagePath = thumbnailPath
            item = xbmcgui.ListItem(convert_timestamp(timestamp=imageDate))
            item.setArt({'icon': thumbnailPath, 'thumb': thumbnailPath})
            contextmenu = []
            contextmenu.append((Const.STR(30010), 'Container.Update(%s)' % build_url({'action': 'search_by_timestamp', 'timestamp': imageDate})))
            if latitude and longitude:
                contextmenu.append((Const.STR(30011), 'Container.Update(%s)' % build_url({'action': 'search_by_latlong', 'latitude': latitude, 'longitude': longitude})))
            contextmenu.append((Const.STR(30014), 'Container.Update(%s,replace)' % build_url({})))
            item.addContextMenuItems(contextmenu, replaceItems=True)
            xbmcplugin.addDirectoryItem(self.handle, imagePath, item, False)
            n += 1
        return n

    def main_menu(self):
        url = build_url({'action': 'moments'})
        item = xbmcgui.ListItem(Const.STR(30001))
        item.setArt({'icon': Const.PICTURE, 'thumb': Const.PICTURE})
        xbmcplugin.addDirectoryItem(self.handle, url, item, True)

        url = build_url({'action': 'people'})
        item = xbmcgui.ListItem(Const.STR(30003))
        item.setArt({'icon': Const.PEOPLE, 'thumb': Const.PEOPLE})
        xbmcplugin.addDirectoryItem(self.handle, url, item, True)

        url = build_url({'action': 'places'})
        item = xbmcgui.ListItem(Const.STR(30002))
        item.setArt({'icon': Const.MARKER, 'thumb': Const.MARKER})
        xbmcplugin.addDirectoryItem(self.handle, url, item, True)

        url = build_url({'action': 'videos'})
        item = xbmcgui.ListItem(Const.STR(30005))
        item.setArt({'icon': Const.MOVIE, 'thumb': Const.MOVIE})
        xbmcplugin.addDirectoryItem(self.handle, url, item, True)

        url = build_url({'action': 'albums', 'folderUuid': 'TopLevelAlbums'})
        item = xbmcgui.ListItem(Const.STR(30004))
        item.setArt({'icon': Const.PHOTO_GALLERY, 'thumb': Const.PHOTO_GALLERY})
        xbmcplugin.addDirectoryItem(self.handle, url, item, True)

        return 4


if __name__ == '__main__':

    args = parse_qs(sys.argv[2][1:])
    for key in args.keys():
        args[key] = args[key][0]

    action = args.get('action', None)
    folderUuid = args.get('folderUuid', None)
    uuid = args.get('uuid', None)
    year = args.get('year', None)
    month = args.get('month', None)
    day = args.get('day', None)

    timestamp = args.get('timestamp', None)
    latitude = args.get('latitude', None)
    longitude = args.get('longitude', None)

    name = args.get('name', None)
    type = args.get('type', None)
    modelId = args.get('modelId', None)

    app = App().open_db()

    if action is None:
        Cache().clear()
        items = app.main_menu()
    elif uuid:
        if type:
            items = app.list_places((name, uuid, int(type), int(modelId)))
            if items == 0:
                items = app.list_photos(uuid, action)
        else:
            items = app.list_photos(uuid, action)
    elif action == 'moments':
        items = app.list_moments(year, month)
    elif action == 'people':
        items = app.list_people()
    elif action == 'places':
        items = app.list_places()
    elif action == 'videos':
        items = app.list_videos()
    elif action == 'albums':
        items = app.list_albums(folderUuid)

    elif action == 'search_by_year':
        items = app.list_photos((year), action)
    elif action == 'search_by_month':
        items = app.list_photos((year, month), action)
    elif action == 'search_by_day':
        items = app.list_photos((year, month, day), action)
    elif action == 'search_by_timestamp':
        items = app.list_photos((timestamp), action)
    elif action == 'search_by_latlong':
        items = app.list_photos((latitude, longitude), action)
    else:
        items = 0

    app.close_db()

    if items == 0:
        notify(Const.STR(30100))
    else:
        xbmcplugin.endOfDirectory(int(sys.argv[1]), True)
